### PR TITLE
[Form] add isSubmittedRequestValid function

### PR DESCRIPTION
**Pull request by Oviglo**
| Q             | A
| ------------- | ---
| Branch?       | 7.1 
| Bug fix?      | no
| New feature?  | yes 
| Deprecations? | no 
| Issues        | 
| License       | MIT

In controllers, we often have to call three functions to validate a form: `handleRequest`, ìsSubmitted`and `isValid`. I wrote a new method that calls `handlerequest` and return the result of the condition `isSubmitted() && isValid()`

### CODE REVIEW

**Review by Arkalo2 — APPROVED** [review]

File: src/Symfony/Component/Form/FormInterface.php [hunk 1]
```diff
@@ -260,6 +260,11 @@ public function initialize(): static;
      */
     public function handleRequest(mixed $request = null): static;
 
+    /**
+     * Call {@link handleRequest()} and return the result of the condition {@link isSubmitted()} and {@link isValid()}
+     */
+    public function isSubmittedRequestValid(mixed $request = null): bool;
```
Comment (yceruto) [inline src/Symfony/Component/Form/FormInterface.php]: Adding a new method to an existent interface is a BC break that is not allowed in minor versions.

You should remove it and add it as a virtual method documented at the top of the interface, so we can finally add it in v8.0

**Review by yceruto — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Form/Form.php [hunk 2]
```diff
@@ -416,6 +416,13 @@ public function handleRequest(mixed $request = null): static
         return $this;
     }
 
+    public function isSubmittedRequestValid(mixed $request = null): bool
```
Comment (yceruto) [inline src/Symfony/Component/Form/Form.php]: I like this method; it simplifies three operations in one call and makes sense to me to relate them as part of the Form component. But, I would typehint `Request` directly here, I don't see any blockers to not doing it; do you?

Comment (Oviglo) [reply src/Symfony/Component/Form/Form.php]: Thanks, I change it, I think $request property don't need to be null.

File: src/Symfony/Component/Form/FormInterface.php (hunk 1 above)
Comment (Oviglo) [reply src/Symfony/Component/Form/FormInterface.php]: You are right, I changed it. Thanks.

File: src/Symfony/Component/Form/FormBuilderInterface.php [hunk 3]
```diff
@@ -15,6 +15,8 @@
  * @author Bernhard Schussek <bschussek@gmail.com>
  *
  * @extends \Traversable<string, FormBuilderInterface>
+ * 
+ * @method bool isSubmittedRequestValid(Request $request) Call {@link handleRequest()} and return the result of the condition {@link isSubmitted()} and {@link isValid()}
```
Comment (yceruto) [inline src/Symfony/Component/Form/FormBuilderInterface.php]: this should be defined over the `FormInterface` instead, right?

File: src/Symfony/Component/Form/Form.php [hunk 4]
```diff
@@ -416,6 +417,13 @@ public function handleRequest(mixed $request = null): static
         return $this;
     }
 
+    public function isSubmittedRequestValid(Request $request): bool
```
Comment (yceruto) [inline src/Symfony/Component/Form/Form.php]: you should keep it as optional (e.i. `?Request $request = null`) as `NativeRequestHandler` will expect a `null` value always

File: src/Symfony/Component/Form/FormBuilderInterface.php (hunk 3 above)
Comment (Oviglo) [reply src/Symfony/Component/Form/FormBuilderInterface.php]: Yes, sorry

**Review by yceruto — APPROVED** [review]
LGTM!

**Comment by yceruto** [comment]
There is a minor CS remaining spotted by fabbot https://fabbot.io/report/symfony/symfony/54726/6b8a8b268f32cd4037a04d217aa8bb551774cf51

**Comment by Oviglo** [comment]
> There is a minor CS remaining spotted by fabbot https://fabbot.io/report/symfony/symfony/54726/6b8a8b268f32cd4037a04d217aa8bb551774cf51

https://fabbot.io/patch/symfony/symfony/54726/6b8a8b268f32cd4037a04d217aa8bb551774cf51/exception_messages.diff
This change request is weird isn't it ? It concern to add '.' before quote in the string

File: src/Symfony/Component/Form/Form.php [hunk 5]
```diff
@@ -301,7 +302,7 @@ public function setData(mixed $modelData): static
             if (null !== $dataClass && !$viewData instanceof $dataClass) {
                 $actualType = get_debug_type($viewData);
 
-                throw new LogicException('The form\'s view data is expected to be a "'.$dataClass.'", but it is a "'.$actualType.'". You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms "'.$actualType.'" to an instance of "'.$dataClass.'".');
+                throw new LogicException('The form\'s view data is expected to be a ".'.$dataClass.'", but it is a "'.$actualType.'". You can avoid this error by setting the "data_class" option to null or by adding a view transformer that transforms "'.$actualType.'" to an instance of "'.$dataClass.'".');
```
Comment (yceruto) [inline src/Symfony/Component/Form/Form.php]: you can ignore this fabbot recommendation, it's not correct.

**Review by yceruto — APPROVED** [review]

**Comment by derrabus** [comment]
I'm not entirely convinced, we need this feature, tbh. My biggest concern is that we change a well established interface and add an unnecessary method to it.

If your main motivation is saving a few keystrokes of boilerplate in controllers, let's add that method to the `AbstractController` class or some trait instead?

**Comment by Oviglo** [comment]
> I'm not entirely convinced, we need this feature, tbh. My biggest concern is that we change a well established interface and add an unnecessary method to it.
> 
> If your main motivation is saving a few keystrokes of boilerplate in controllers, let's add that method to the `AbstractController` class or some trait instead?

This is an interesting observation, I can make a new PR to write this in AbstractController:
```php
$form = $this->createForm(FormType::class);
if ($this->isSubmittedFormValid($form, $request)) {
}
```
I don't think this will be useful except to not forget the `handleRequest` method

File: src/Symfony/Component/Form/Form.php [hunk 6]
```diff
@@ -416,6 +417,13 @@ public function handleRequest(mixed $request = null): static
         return $this;
     }
 
+    public function isSubmittedRequestValid(?Request $request = null): bool
```
Comment (derrabus) [inline src/Symfony/Component/Form/Form.php:420]: This signature breaks the component's abstraction, btw: `handleRequest()` has no opinion on how a request is represented, but this method surprisingly requires an HttpFoundation request.

**Review by derrabus — CHANGES_REQUESTED** [review]

**Comment by derrabus** [comment]
> This is an interesting observation, I can make a new PR to write this in AbstractController:
> 
> ```
> $form = $this->createForm(FormType::class);
> if ($this->isSubmittedFormValid($form, $request)) {
> }
> ```

If we want to have such a utility function in the core at all, I think `AbstractController` is the place.

File: src/Symfony/Component/Form/Form.php (hunk 6 above)
Comment (Oviglo) [reply src/Symfony/Component/Form/Form.php:420]: I changed this at @yceruto 's request: (cf: https://github.com/symfony/symfony/pull/54726#discussion_r1579731842)

**Comment by Oviglo** [comment]
It rewrite the method in AbstractController: https://github.com/symfony/symfony/pull/54743

**Pull request closed without merging**